### PR TITLE
DEMO: Testing framework

### PR DESCRIPTION
**Pull request by jpeletier**
This PR is used as a demo of a proposed testing framework for Swarm. In the PR, I have partly refactored some of Feeds tests to illustrate the code savings and improvement in readability of the test code, and how test results move to the `testdata/` directory.

Although this PR could be merged, the goal is to study the diff and see for yourself the number of lines of code saved and the potential time saved when writing test code.

The structure of the typical test now becomes:
```go
func TestExample (tx * testing.T) {
	t := testutil.BeginTest(tx, false) // set to true to generate results
	defer t.FinishTest() // this will cleanup automatically

	// obtain a temporary dir so SomeFunction can write stuff there
	tempDir := t.Services.NewTempDir()

	response, err := somepackage.SomeFunction(tempDir)
	t.Ok(err)
	t.EqualsKey("response", response) //this compares the response with the recorded answer in /testdata/results.json, which is generated automatically when you set the BeginTest parameter to true

	_, err = somepackage.BadFunc()
	t.MustFail(err, "Expected BadFunc to return an error, since it is such a bad func")
	// etc 
}
```

... and much more. You have `Assert()`, `Equals()`, `JSONEquals()`...

### CODE REVIEW

**Comment by jpeletier** [comment]
Documentation available here: https://github.com/epiclabs-io/ut

File: swarm/storage/feed/lookup/epoch_test.go [hunk 1]
```diff
@@ -1,34 +1,40 @@
...
 		e := lookup.Epoch{
 			Time:  i,
 			Level: uint8(i % 20),
 		}
 		b, err := e.MarshalBinary()
-		if err != nil {
-			t.Fatal(err)
-		}
+		t.Ok(err)
+		t.EqualsKey(fmt.Sprintf("epoch%018d", i), hexutil.Bytes(b))
```
Comment (nonsense) [inline swarm/storage/feed/lookup/epoch_test.go:23]: @jpeletier it is not obvious to a new-comer what this is doing. I suggest renaming this to something related to `golden` files - I think this is the most common name for such kind of tests in Go.

One post documenting them is https://medium.com/soon-london/testing-with-golden-files-in-go-7fccc71c43d3

Maybe: `t.CompareKeyWithGoldenFile` ? Although that is a bit long... to me `t.EqualsKey` suggests that you are comparing the two parameters.

File: swarm/storage/feed/request_test.go [hunk 2]
```diff
@@ -271,42 +212,30 @@ func TestReverse(t *testing.T) {
...
+	t.Ok(err)
+
 	originalAddr := crypto.PubkeyToAddress(signer.PrivKey.PublicKey)
 
 	// check that the metadata retrieved from the chunk matches what we gave it
-	if recoveredAddr != originalAddr {
-		t.Fatalf("addresses dont match: %x != %x", originalAddr, recoveredAddr)
-	}
+	t.Assert(recoveredAddr == originalAddr, "addresses dont match: %x != %x", originalAddr, recoveredAddr)
+	t.Equals(chunk.Address()[:], key[:])
```
Comment (nonsense) [inline swarm/storage/feed/request_test.go:236]: One of the gotchas of the `t.Equals` that we would have to be very careful with:

```
	a := 1
	b := 1

	t.Equals(a, b)
	t.Equals(&a, &b)

	if a != b {
		t.Fatal("a != b")
	}

	if &a != &b {
		t.Fatal("&a != &b")
	}
```

This is one of the reasons I don't like frameworks like `testify`.

Comment (nonsense) [reply swarm/storage/feed/request_test.go:236]: @jpeletier as explained during the call, types are not always compared as you would expect.

**Review by lmars — COMMENTED** [review]
I think it's a great idea to provide more clarity in tests because they can very easily become unwieldy, and I understand the desire to add things like `t.Ok` and `t.Assert`, but I think the solution is to actually provide many more specific helper functions in tests, so that the body of the test is a little smaller and easy to see what is being tested, and the helper functions do a lot of the heavy lifting.

For example, imagine if `TestFeedsHandler` looked something like:

```go
func TestFeedsHandler(t *testing.T) {
	x := testFeedsHelper()

	// create a fake clock
	clock := x.newFakeClock(t)

	// create a signer
	signer := x.newTestSigner(t)

	// create a test feed
	feed := x.createTestFeed(t, signer)

	// define some data updates
	updates := []string{
		"blinky", // t=4200
		"pinky",  // t=4242
		"inky",   // t=4284
		"clyde",  // t=4285
	}

	// update the feed with updates[0]
	feed.update(t, updates[0])

	// move the clock ahead 21 seconds
	clock.FastForward(21 * time.Second)

	// create a new request which timestamps the update at t = 4221
	request := feed.newRequest(t)

	// check that request.Epoch has Base 0 and Level 1 less than the highest level
	if request.Epoch.Base() != 0 || request.Epoch.Level != lookup.HighestLevel-1 {
		t.Fatalf("Suggested epoch BaseTime should be 0 and Epoch level should be %d", lookup.HighestLevel-1)
	}

	// etc...
}
```

**Comment by zelig** [comment]
closing as team sees not enough added value for the complexity

**Pull request closed without merging**